### PR TITLE
Implement Clear-SbQueue (Purge)

### PR DESCRIPTION
**Pull request by Splaxi**

### CODE REVIEW

**Comment by Splaxi** [comment]
6K messages was purged in less than 60 seconds. I'm totally hooked on this! 🤘

**Comment by tmeadon** [comment]
That's awesome!  Have you experimented with increasing the number of messages to receive per call to see if receiving is limited by that AMQP mass transit limit?  Have you also checked out what happens if use prefetch?

I'm wondering if we can get it _even_ faster! 😉

**Comment by Splaxi** [comment]
The original code was actually with prefetch, so I'll be adding that again.

I'm testing things to see when it breaks and will also test whether the 100 messages are enforced or not...

**Comment by tmeadon** [comment]
Awesome! If you do find a breaking point then maybe we should add a test to ensure that we never exceed that point.

**Comment by tmeadon** [comment]
The code for this is looking good by the way.  Let me know when it is ready for a final review before merge!

**Comment by Splaxi** [comment]
This was a CRAZY issue that I hit with this.

The whole thing about the task / await pattern and not understanding how things are connected, threw me way of. I hope that I have fixed it for good now.

File: src/PSServiceBus/Helpers/SbReceiver.cs [hunk 1]
```diff
@@ -167,13 +173,27 @@ public void PurgeMessages()
...
+                        break;
+                    }
+
+                    //Is the messageReceiver in a closing state?
+                    if (messageReceiver.IsClosedOrClosing)
+                    {
+                        break;
+                    }
+
+                    var temp = messageReceiver.ReceiveAsync(100, TimeSpan.FromSeconds(1));
```
Comment (tmeadon) [inline src/PSServiceBus/Helpers/SbReceiver.cs:188]: For my own knowledge - what is the difference between these 3 lines and

```
res = messageReceiver.ReceiveAsync(100, TimeSpan.FromSeconds(1)).Result;
```

**Review by tmeadon — APPROVED** [review]

**Comment by tmeadon** [comment]
This is looking good @Splaxi - ready to merge?

**Comment by tmeadon** [comment]
Oh I should add, the additions you made to support disposal on messageReceiver are pushing my C# knowledge so I'm not in a great place to review them, but, what's the worst that happen!?  😁 

Quick question - if your latest commits are fixing some kind of bug, should we add a test to expose the bug to make sure we don't re-introduce it later?

File: src/PSServiceBus/Helpers/SbReceiver.cs (hunk 1 above)
Comment (Splaxi) [reply src/PSServiceBus/Helpers/SbReceiver.cs:188]: I'm "hoping" that the .Wait() will actually make things wait and not run it in some kind of back ground thread (async) mode.

When reading about the task / async pattern, you will see it is very hard to make synchronous code execute asynchronous code, without things running in the back ground. This is my best guess with my current knowledge, and my tests are showing it works.

**Comment by Splaxi** [comment]
Let's just say my limits were also pushed with this issue. As long as MS don't support a real canceltoken approach, this seems to be the best way to "fake" the same behavior.

The real issue was that the receiver was still open and waiting (in the background) for new messages. So testing this would require us to pump messages (1000's) into the queue, purge them, pump messages (1000's) into the queue again and then see if the ActiveMessageCount matches. This was how I discovered the issue, while testing my own workload and missing exactly 100 messages. Look and behold it matched the Prefetch = 100, but it took me days to realize that.

**Comment by tmeadon** [comment]
Ahhh that does sound like a bit of a nasty issue...  Why would we need to pump 1000s of messages in to test this?  Wouldn't that background prefetch issue surface up when using any number of messages > the prefetch count?

**Comment by Splaxi** [comment]
I had the same logic in the beginning, but I was only seeing this error when I was working with 1000s of messages. Or maybe I simple tried so darn many things over the course of the different evenings, that I lost some details along the way.

Let us get this into the module and then I'll start testing it at bit more, to come up with at integration tests that proves things are working.

**Comment by tmeadon** [comment]
Sounds good, I'm happy with that approach.  Since adding those tests isn't a pressing matter, you could raise an issue that we can come back to down the line if you want to work on something else next!

**Pull request merged**